File: TwoPointers/RemoveNthLastNode/RemoveNthLastNode.py


class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

def removeNthFromEnd(head, n):
    if head is None:
        return None

    tail = head
    delete = head
    for _ in range(n):
        tail = tail.next
    if (tail == None):
        return head.next
    while(tail.next):
        tail = tail.next
        delete = delete.next
    delete.next = delete.next.next
    return head

h1 = ListNode(1)
h2 = ListNode(2)
h3 = ListNode(3)
h4 = ListNode(4)
h1.next = h2
h2.next = h3
h3.next = h4
res = removeNthFromEnd(h1,4)
print(res.val if res else None)

File: RemoveNthLastNode/RemoveNthLastNode.java
public class RemoveNthLastNode {
    public static class ListNode {
        int val;
        ListNode next;
        ListNode(int val) { this.val = val; }
        ListNode(int val, ListNode next) { this.val = val; this.next = next; }
    }
    public ListNode removeNthFromEnd(ListNode head, int n) {
        ListNode tail = head;
        ListNode del = head;
        for(int i = 0; i < n; i++){
            tail = tail.next;
        }
        if (tail == null)
            return head.next;
        while(tail.next != null){
            tail = tail.next;
            del = del.next;
        }
        del.next = del.next.next;
        return head;
    }

    public static void main(String[] args) {
        ListNode head = new ListNode(1,
                new ListNode(2,
                        new ListNode(3,
                                new ListNode(4,
                                        new ListNode(5)))));


        RemoveNthLastNode solution = new RemoveNthLastNode();
        ListNode result = solution.removeNthFromEnd(head, 5);
        if (result != null){
            System.out.println(result.val);
        }

        
    }

    
}
